modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts

```diff
@@ -0,0 +1,207 @@
+  if (a.tag === 'JUST' && b.tag === 'JUST' && c.tag === 'JUST' && d.tag === 'JUST') {
+    return just(f(a.value, b.value, c.value, d.value));
+  } else {
+    return nothing;
+  }
+};
+
+export const lift5 = <A, B, C, D, E, Out>(a: Maybe<A>, b: Maybe<B>, c: Maybe<C>, d: Maybe<D>, e: Maybe<E>, f: (a: A, b: B, c: C, d: D, e: E) => Out): Maybe<Out> => {
+  if (a.tag === 'JUST' && b.tag === 'JUST' && c.tag === 'JUST' && d.tag === 'JUST' && e.tag === 'JUST') {
+    return just(f(a.value, b.value, c.value, d.value, e.value));
+  } else {
+    return nothing;
+  }
+};
+
+// Monad
+export const bind = <T, U>(self: Maybe<T>, f: (input: T) => Maybe<U>): Maybe<U> => {
+  switch (self.tag) {
+    case 'JUST':
+      return f(self.value);
+    case 'NOTHING':
+      return self;
+  }
+};
+
+export const bind2 = <T, U>(self: Maybe<T>, f: (input: T) => U | null | undefined): Maybe<NonNullable<U>> =>
+  bind(self, (inner) => from(f(inner)));
+
+export const flatten = <T>(self: Maybe<Maybe<T>>): Maybe<T> => {
+  switch (self.tag) {
+    case 'JUST':
+      return self.value;
+    case 'NOTHING':
+      return self;
+  }
+};
+
+// Side effects
+export const each = <T>(self: Maybe<T>, f: (input: T) => void): void => {
+  switch (self.tag) {
+    case 'JUST':
+      f(self.value);
+      break;
+    case 'NOTHING':
+      // do nothing
+  }
+};
+
+// Unwrapping
+export const or = <T>(self: Maybe<T>, other: Maybe<T>): Maybe<T> => {
+  switch (self.tag) {
+    case 'JUST':
+      return self;
+    case 'NOTHING':
+      return other;
+  }
+};
+
+export const orThunk = <T>(self: Maybe<T>, thunk: () => Maybe<T>): Maybe<T> => {
+  switch (self.tag) {
+    case 'JUST':
+      return self;
+    case 'NOTHING':
+      return thunk();
+  }
+};
+
+export const getOr = <T>(self: Maybe<T>, other: T): T => {
+  switch (self.tag) {
+    case 'JUST':
+      return self.value;
+    case 'NOTHING':
+      return other;
+  }
+};
+
+export const getOrDie = <T>(self: Maybe<T>): T => {
+  switch (self.tag) {
+    case 'JUST':
+      return self.value;
+    case 'NOTHING':
+      throw new Error('Called getOrDie on Nothing');
+  }
+};
+
+export const getOrThunk = <T>(self: Maybe<T>, thunk: () => T): T => {
+  switch (self.tag) {
+    case 'JUST':
+      return self.value;
+    case 'NOTHING':
+      return thunk();
+  }
+};
+
+export const getOrNull = <T>(self: Maybe<T>): T | null =>
+  getOr(self, null);
+
+export const getOrUndefined = <T>(self: Maybe<T>): T | undefined =>
+  getOr(self, undefined);
+
+// Foldable / array style things
+export const exists = <T>(self: Maybe<T>, pred: (input: T) => boolean): boolean =>
+  self.tag === 'JUST' && pred(self.value);
+
+export const forAll = <T>(self: Maybe<T>, pred: (input: T) => boolean): boolean =>
+  self.tag === 'NOTHING' || pred(self.value);
+
+export const filter: {
+  <T>(self: Maybe<T>, pred: (input: T) => boolean): Maybe<T>;
+  <T, U extends T>(self: Maybe<T>, pred: (input: T) => input is U): Maybe<U>;
```

Comment:
I thought typescript tried applying types top to bottom, in which case this would not gain the type filtering. Have you checked this applies the type filtering?

Reply:
Good catch: this is fixed